org.eclipse.scanning.api/src/org/eclipse/scanning/api/database/CompositeBean.java

```diff
@@ -21,55 +26,86 @@
  *
  */
 public class CompositeBean {
+	
+	private static final Logger logger = LoggerFactory.getLogger(CompositeBean.class);
 
 	/**
 	 * IMPORTANT: Do not use generics here, the data
 	 * is almost always heterogeneous.
 	 */
-	private Map<Operation, Collection<Object>> data;
+	private Map<Operation, Collection<Bean>> data;
     
     public CompositeBean() {
     	data = new LinkedHashMap<>(); // Order matters in the collection of items
     }
     
-    public CompositeBean(Operation type, Object... beans) {
+    /**
+     * 
+     * @param type
+     * @param originalObjectsOrBeans may be some Bean objects or the actual database objects (which will be wrapped with Bean)
+     */
+    public CompositeBean(Operation type, Object... originalObjectsOrBeans) {
     	this();
     	assertType(type);
-     	data.put(type, new ArrayList<>(Arrays.asList(beans)));
+  
+    	List<Bean> wrapped = new ArrayList<Bean>(originalObjectsOrBeans.length);
+    	Arrays.asList(originalObjectsOrBeans).forEach(object->wrapped.add(toBean(object)));
+     	data.put(type, wrapped);
     }
    
-    private void assertType(Operation type) {
+    private Bean toBean(Object object) {
+    	try {
+			return object instanceof Bean ? (Bean)object : new Bean(object);
+		} catch (NoSuchMethodException 
+					| SecurityException 
+					| IllegalAccessException 
+					| IllegalArgumentException
+					| InvocationTargetException e) {
+			
+			logger.error("Unexpected object tried to be added to "+getClass().getSimpleName(), e);
+			return null;
+		}
+    }
+
+	private void assertType(Operation type) {
        	if (!Operation.ends().contains(type)) {
        		throw new IllegalArgumentException("Each operation set added to the composite must not be a composite. Allowed types are "+Operation.ends());
        	}
 	}
 
-	public void add(Operation type, Object bean) {
+	public void add(Operation type, Bean bean) {
     	
     	assertType(type);
-    	Collection<Object> beans = data.get(type);
+    	Collection<Bean> beans = data.get(type);
     	if (beans==null) {
     		beans = new ArrayList<>(7); // Order matters in the collection of items
     		data.put(type, beans);
     	}
     	beans.add(bean);
     }
     
-    public Collection<Object> set(Operation type, Object... beans) {
+	/**
+	 * 
+	 * @param type
+     * @param originalObjectsOrBeans may be some Bean objects or the actual database objects (which will be wrapped with Bean)
+	 * @return
+	 */
+    public Collection<Bean> set(Operation type, Object... originalObjectsOrBeans) {
        	assertType(type);
-    	return data.put(type, new ArrayList<>(Arrays.asList(beans)));
+    	List<Bean> wrapped = new ArrayList<Bean>(originalObjectsOrBeans.length);

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2293)